feat: Add `rust_lint_group` and `cargo_lints` rule to define lint groups

ParkMyCar opened the pull request:
Chatted about this previously in [Slack](https://bazelbuild.slack.com/archives/CSV56UT0F/p1718030356910019). The goal with this PR is to make it easier to define and share lint configurations for your build.

### Description

The PR adds two new rules, `rust_lint_group` and `cargo_lints`. It also adds a "lints" attr to `rust_library` and `rust_binary` rules that expects a new `LintsInfo` provider.

* `rust_lint_group` allows you to define Rust, Clippy, and Rustdoc lints in a `BUILD` file.
* `cargo_lints` automatically generates a set of lints by reading a crate's `Cargo.toml`, and optionally the workspace's Cargo.toml for workspace inheritance.

Then the rustc, clippy, and rustdoc actions are all updated to check for the `LintsInfo` provider and appends the correct arguments so the lints take effect.


### Design

Honestly this is my first large change to a set of Bazel rules, so I definitely could have done something wrong here!

The change is split into two commits, the first introduces `rust_lint_group` which IMO is relatively straight forward. Some attributes are defined on a rule which are then formatted into command line flags and passed around with a provider.

The second commit adds `cargo_lints` and is much larger, a few things worth considering:

* I was back and forth a bit on using a `repository_rule` or a regular `rule`. While the `repository_rule` maps well to a Cargo Workspace and you'd only need to define it once, not everyone uses workspaces and so I figured a regular `rule` was more general.
* Parsing a `Cargo.toml` is done via a new Rust binary called `cargo_toml_info`. I tried to keep the external dependencies on this binary to a minimum, it only has one at the moment which is [`cargo_toml`](https://docs.rs/cargo_toml/latest/cargo_toml/) and I based this change largely off of https://github.com/bazelbuild/rules_rust/pull/2772. I tried to make the tool general though so other Cargo metadata rules could use it in the future.

### Tests

There aren't any! I wasn't sure where the best place to start was, any guidance here is appreciated!

illicitonion commented on cargo/cargo_toml_info/main.rs:
```diff
@@ -0,0 +1,272 @@
+    output_rustc_lints: PathBuf,
+    output_clippy_lints: PathBuf,
+    output_rustdoc_lints: PathBuf,
+    // TODO(parkmycar): Support Rust's new --check-cfg once cargo-toml is updated.
```
Can we link to an issue?

ParkMyCar replied:
Filed and linked to [cargo_toml#34](https://gitlab.com/lib.rs/cargo_toml/-/issues/34)!

illicitonion commented on cargo/private/cargo_lints.bzl:
```diff
@@ -0,0 +1,73 @@
+        )
+    ]
+
+cargo_lints = rule(
```
Maybe `extract_cargo_lints`?

ParkMyCar replied:
Done!

illicitonion reviewed (commented):
This generally looks super reasonable, thanks so much for putting it together!

I just have a few comments which are basically all naming.

In terms of testing, one problem we have in Bazel in general is it's pretty hard to write tests that things which should fail do fail (particularly without them being wildly slow).

There are generally three styles of tests you can probably write for this functionality:

I would recommend mostly writing unit tests - you can see an example of how we write unit tests in https://github.com/bazelbuild/rules_rust/tree/main/test/unit/extra_rustc_flags - your test would probably look pretty similar, just testing that some extra flags are present.

For the rust code, you can just write some unit tests like normal.

And you can write a script that checks failing things do fail - we have an example for clippy that runs on CI: https://github.com/bazelbuild/rules_rust/blob/60515a1defef5496ac0f5e17cff6404d2d2feeeb/.bazelci/presubmit.yml#L367-L371 - these ones definitely don't have to be exhaustive, just one example showing each style of thing can fail (and ideally fails with a useful error message) is probably sufficient :)

ParkMyCar commented on rust/private/providers.bzl:
```diff
@@ -174,3 +174,12 @@ RustAnalyzerGroupInfo = provider(
+    fields = {
+        "rustc_lints": "List[String]: rustc flags to specify when building rust_* targets.",
+        "clippy_lints": "List[String]: rustc flags to specify when building clippy targets.",
+        "rustdoc_lints": "List[String]: rustc flags to specify when building rust_doc targets.",
```
Good idea, done!

ParkMyCar commented on rust/defs.bzl:
```diff
@@ -157,4 +161,7 @@ rustfmt_test = _rustfmt_test
 rust_stdlib_filegroup = _rust_stdlib_filegroup
 # See @rules_rust//rust:toolchain.bzl for a complete description.
 
+rust_lint_group = _rust_lint_group
```
Yeah IMO "config" is definitely better than "group", changed!

ParkMyCar commented on cargo/cargo_toml_info/main.rs:
```diff
@@ -0,0 +1,272 @@
+//! Command line tool invoked by Bazel to read metadata out of a `Cargo.toml` file.
```
Done! Also repinned all dependencies which created a number of changes, sorry for the noise in the commit

ParkMyCar commented on cargo/cargo_toml_info/main.rs:
```diff
@@ -0,0 +1,272 @@
+        let manifest = Manifest::from_path(&workspace_path)?;
+        let workspace_details = Some((&manifest, workspace_path.as_path()));
+
+        // TODO(parkmycar): Fix cargo_toml so we inherit lints from our workspace.
```
In the meantime we "manually" inherit lints by checking both the manifest and workspace `Cargo.toml` so the behavior is correct, but ideally this would be handled internally by `cargo_toml`. Filed an [issue](https://gitlab.com/lib.rs/cargo_toml/-/issues/35) upstream and left a comment here

ParkMyCar commented:
Thanks for the review @illicitonion! Didn't get a chance to add tests (although I did test manually by using this in my project), will follow up with that!

ParkMyCar commented:
Had a free cycle so I added an initial unit test for `rust_lint_group` more than happy to add more when I have some time!

ParkMyCar commented:
@illicitonion whenever you have some time (no rush!) this is ready for a re-review! Sorry for all the commits, there were a bunch of tiny tweaks to get CI green. Overall you can probably skip the first 5 commits, nothing there changed since your last review.

What changed since your last review is:
1. Added bzlmod support for `extract_cargo_lints` by including the relevant repositories in `rules_rust`'s root `MODULE.bazel`.
2. Adjusted the flags set for Clippy to also include rustc lints, this mimics what Cargo does.
3. Added a unit test to exercise the `rust_lint_config` rule, in `test/unit/lint_flags`
4. Added an example to exercise the `extract_cargo_lints` rule, in `examples/cargo_lints`
5. Updated an existing example to make sure `extract_cargo_lints` worked in bzlmod, in `examples/bazel_env`.

illicitonion commented on cargo/private/cargo_toml_info/main.rs:
```diff
@@ -0,0 +1,276 @@
+    /// Expects 4 filesystem paths in this order:
+    ///
+    /// 1. output for rustc lints
+    /// 2. output for rust check-cfg lints
```
We should probably remove this expected arg, enum variant, and such until we actually support it. WDYT?

ParkMyCar replied:
Makes sense to me! Removed all references to it in the `cargo_lints` rule in the latest commit.

illicitonion commented on examples/cargo_lints/Cargo.lock:
I don't think we need `Cargo.lock` files for examples where all we're doing is parsing the `Cargo.toml` - can you remove and .gitignore?

ParkMyCar replied:
Removed!

illicitonion reviewed (commented):
Sorry it's taken so long, this looks good to me! Let's rebase, fix up the couple of tiny things, and get this merged!

ParkMyCar commented:
> Sorry it's taken so long, this looks good to me! Let's rebase, fix up the couple of tiny things, and get this merged!

No problem at all! I appreciate you working through this one with me :)

ParkMyCar commented:
Friendly ping @illicitonion so this can try and make the v0.57 release :)

illicitonion reviewed (approved):
Thanks!
